File: AdventOfCode2021/d20_nim/d20.nim
import std/strutils
import std/tables
import std/sets
import std/sequtils

type Point = tuple[x: int, y: int]
let mvs = [(1, 1), (1, 0), (1, -1), (0, 1), (0, 0), (0, -1), (-1, 1), (-1, 0), (-1, -1)]

proc ns(pt: Point): seq[Point] =
    var o: seq[Point] = @[]
    for mv in mvs:
        o.add((pt.x+mv[0], pt.y+mv[1]))
    return o

let lines = splitLines(readAll(stdin))

let alg = lines[0]

let ob = newTable[Point, char]()

for x, line in lines[2 .. ^1]:
    for y, c in line:
        ob[(x, y)] = c

proc pts2Update(tbl: TableRef[Point, char]): seq[Point] =
    var s: HashSet[Point] = initHashSet[Point]()
    for k in tbl.keys:
        for n in ns(k):
            s.incl(n)
    return toSeq(s)

proc sim(its: int): int =
    var b = ob
    var rest: char = '.'
    for i in 1 .. its:
        var nb = newTable[Point, char]()
        for pt in pts2Update(b):
            var sm = 0
            for off, n in ns(pt):
                if b.getOrDefault(n, rest) == '.':
                    continue
                sm += 1 shl off
            nb[pt] = alg[sm]
        if rest == '.':
            rest = alg[0]
        else:
            rest = alg[^1]
        b = nb
    var cnt = 0
    for v in b.values:
        if v == '#':
            cnt += 1
    return cnt
    
echo "Part 1: ", sim(2)
echo "Part 2: ", sim(50)

File: AdventOfCode2021/pysolns/d20.py
import sys
from itertools import product
from collections import defaultdict

mvs = [(1, 1), (1, 0), (1, -1), (0, 1), (0, 0), (0, -1), (-1, 1), (-1, 0), (-1, -1)] 

def lines():
  return [line.strip() for line in sys.stdin]

def ns(pt):
  return [(n[0] + pt[0], n[1] + pt[1]) for n in mvs]

l = lines()
alg = l[0]

ob = [[x for x in l] for l in l[2:]]
ob = {(x, y): ob[x][y] for x, y in product(range(len(ob)), range(len(ob[0])))}

def sim(its):
  b = ob
  rest = '.'
  for _ in range(its):
    pts2up = list(set([n for pt in b.keys() for n in ns(pt)]))
    nb = {}
    for pt in pts2up:
      sm = 0
      for off, n in enumerate(ns(pt)):
        v = b[n] if n in b else rest
        if v == '.':
          continue
        sm += 2**off
      nb[pt] = alg[sm]
    rest = alg[0] if rest == '.' else alg[2**8 - 1]
    b = nb
  return len([k for k,v in b.items() if v == '#'])

def partA():
  print('Part 1:', sim(2))

def partB():
  print('Part 2:', sim(50))

partA()
partB()
